src/boxy.nim

```diff
@@ -361,121 +343,187 @@ proc grow(boxy: Boxy) =
   var imageInfo: ImageInfo
-  imageInfo.size = ivec2(image.width.int32, image.height.int32)
+  var reusing = false
+  if key in boxy.entries:
+    imageInfo = boxy.entries[key]
+    if imageInfo.cap.x >= image.width and imageInfo.cap.y >= image.height:
+      reusing = true
+
+  if not reusing:
+    # New image can't fit in the existing image info.
+    # Remove the existing image info and create a new one.
+    boxy.removeImage(key)
+    imageInfo = ImageInfo()
+    boxy.entriesBuffered.incl(key)
+    imageInfo.size = ivec2(image.width.int32, image.height.int32)
+    imageInfo.cap = imageInfo.size
+    imageInfo.isOneColor = false
+
+    # Try to pack the image using the allocator
+    var packed = false
+    var x, y: int
+    while not packed:
+      let allocation = boxy.allocator.allocate(image.width, image.height)
+      if allocation.success:
+        x = allocation.x
+        y = allocation.y
+        packed = true
+      else:
+        # Need to grow the atlas
+        boxy.grow()
+    imageInfo.atlasPos = ivec2(x.int32, y.int32)
+
+    # Update the image info size
+    imageInfo.size = ivec2(image.width.int32, image.height.int32)
+    # Copy the image to the atlas at the packed position
+    updateSubImage(
+      boxy.atlasTexture,
+      imageInfo.atlasPos.x,
+      imageInfo.atlasPos.y,
+      image
+    )
+
+    boxy.entries[key] = imageInfo
 
-  if image.isOneColor():
-    imageInfo.oneColor = image[0, 0].color
   else:
-    var
-      image = image
-      level = 0
-    while true:
-      imageInfo.tiles.add(@[])
-
-      # Split the image into tiles.
-      for y in 0 ..< boxy.tileHeight(image.height):
-        for x in 0 ..< boxy.tileWidth(image.width):
-          let tileImage = image.superImage(
-            x * boxy.tileSize - tileMargin div 2,
-            y * boxy.tileSize - tileMargin div 2,
-            boxy.tileSize + tileMargin,
-            boxy.tileSize + tileMargin
-          )
-          if tileImage.isOneColor():
-            let tileColor = tileImage[0, 0].color
-            imageInfo.tiles[level].add(
-              TileInfo(kind: tkColor, color: tileColor)
-            )
-          else:
-            let index = boxy.takeFreeTile()
-            imageInfo.tiles[level].add(TileInfo(kind: tkIndex, index: index))
-            updateSubImage(
-              boxy.atlasTexture,
-              (index mod boxy.tileRun) * (boxy.tileSize + tileMargin),
-              (index div boxy.tileRun) * (boxy.tileSize + tileMargin),
-              tileImage
-            )
-
-      if image.width <= 1 or image.height <= 1:
-        break
-
-      if not genMipmaps:
-        break
-
-      image = image.minifyBy2()
-      inc level
-
-  boxy.entries[key] = imageInfo
+    # Update the image info size
+    imageInfo.size = ivec2(image.width.int32, image.height.int32)
+    if imageInfo.cap != imageInfo.size:
+      # Got to clear the margin around the image.
+      boxy.atlasTexture.clearSubImage(
```

Comment:
Could have another path for this, generating only the geo for the margin; saves on filling the entire center which is the most pixels this consumes

Reply:
Or better keep a margin mesh ready to go, and scale it in the vertex shader